baselines/ppo2/policies.py

```diff
@@ -0,0 +1,207 @@
+        }
+        if self.states:
+            self.step_input.update({'states': self.states['current']})
+            self.step_output.update({'states': self.states['current'],
+                                     'next_states': self.states['next']})
+
+    def feed_dict(self, **kwargs):
+        feed_dict = {}
+        for key in kwargs:
+            if key in self.step_input:
+                feed_dict[self.step_input[key]] = adjust_shape(self.step_input[key], kwargs[key])
+        return feed_dict
+
+    def step(self, **kwargs):
+        return self.sess.run(self.step_output,
+                             feed_dict=self.feed_dict(**kwargs))
+
+    def values(self, **kwargs):
+        return self.sess.run({'values': self.value},
+                             feed_dict=self.feed_dict(**kwargs))
+
+    def save(self, save_path):
+        tf_util.save_state(save_path, sess=self.sess)
+
+    def load(self, load_path):
+        tf_util.load_state(load_path, sess=self.sess)
+
+
+def build_ppo_policy(env, policy_network, value_network=None, estimate_q=False, **policy_kwargs):
+    if isinstance(policy_network, str):
+        network_type = policy_network
+        policy_network = get_network_builder(network_type)(**policy_kwargs)
+
+    if value_network is None:
+        value_network = 'shared'
+
+    def policy_fn(nbatch=None, nsteps=None, sess=None, observ_placeholder=None):
+        next_states_list = []
+        state_map = {}
+        value_state = None
+        policy_state = None
+        state_placeholder = None
+
+        ob_space = env.observation_space
+        X = observ_placeholder if observ_placeholder is not None else observation_placeholder(ob_space,
+                                                                                              batch_size=nbatch)
+        dones = tf.placeholder_with_default(np.zeros([X.shape[0]]), [X.shape[0]], name='dones')
+
+        encoded_x = encode_observation(ob_space, X)
+
+        with tf.variable_scope('load_rnn_memory'):
+            if is_rnn_network(policy_network):
+                policy_state, policy_network_ = policy_network(encoded_x, dones)
+            else:
+                policy_network_ = policy_network
+
+            if value_network == 'shared':
+                value_network_ = value_network
+            else:
+                if value_network == 'copy':
+                    value_network_ = policy_network
+                else:
+                    assert callable(value_network)
+                    value_network_ = value_network
+
+                if is_rnn_network(value_network_):
+                    value_state, value_network_ = value_network_(encoded_x, dones)
+
+            if policy_state or value_state:
+                states_list = [state for state in [policy_state, value_state] if state]
+                states = tf.concat(states_list, axis=1)
+                state_placeholder = tf.placeholder_with_default(states, states.get_shape())
```

Comment:
If I understand correctly, this placeholder is not actually being returned anywhere outside of this function (it is stores in state_map, but then state_map is not being returned / referred to outside either). What is the point of making it a placeholder (i.e. how would external code feed data into it)?

Reply:
It will be removed.

Reply:
removed at b6e6c52